R/shapleyR.R

```diff
@@ -51,8 +51,8 @@ shapley = function(row.nr, model = train("regr.lm", bh.task), task = bh.task,
       b2[s:(s + nrow(x) - 1), perm] = cbind(x[prec], z[feature], z[succ])
     }
 
-    phi[feature] = getPredictionResponse(predict(model, newdata=b1)) -
-      getPredictionResponse(predict(model, newdata = b2))
+    phi[feature] = as.numeric(getPredictionResponse(predict(model, newdata=b1))) -
+      as.numeric(getPredictionResponse(predict(model, newdata = b2)))

```

Comment:
Leider funktioniert das nicht so einfach...

`getPredictionResponse(predict(model, newdata=b1)` gibt dir die erwartete Klasse zurück:
> getPredictionResponse(predict(model, newdata = b2))
 [1] setosa     setosa     setosa     setosa     virginica  setosa     setosa     setosa     setosa     virginica  setosa    
[12] setosa     setosa     virginica  setosa     setosa     setosa     setosa     setosa     setosa     setosa ...

Durch das as.numeric wandelst du es dann in 1 bis 3 um (da wir 3 Klassen haben). Bei zwei Klassen könnte es noch Sinn machen, aber bei 3 und mehr Klassen nicht mehr, da 1,2,3 nur Kodierungen sind und keine Information über eine Ordnung geschweige denn Distanz beinhalten. Aber genau das stellt dann der Algorithmus dar, wenn du die Predictions von b1 und b2 subtrahierst.

Reply:
Dann macht es bei Klassifizierungen also nur Sinn zu überprüfen, ob die prediction response von b1 sich zu b2 unterscheidet. Daraus lässt sich auch der shapley wert berechnen. Wie das geht habe ich vor ein paar tagen gelesen.